src/include/OSL/rendererservices.h

```diff
@@ -29,6 +30,125 @@ typedef const void* TransformationPtr;
 typedef void (*PrepareClosureFunc)(RendererServices*, int id, void* data);
 typedef void (*SetupClosureFunc)(RendererServices*, int id, void* data);
 
+enum class AttributeSpecBuiltinArg {
+    ShaderGlobalsPointer,  // void* (ideally ShaderGlobals*)
+    ShadeIndex,            // int
+    Derivatives,           // int (ideally bool)
+    Type,                  // long long (ideally TypeDesc)
+    ArrayIndex,            // int, Always zero for non-indexed array lookups.
+    ArrayLookup,           // int (ideally bool)
+    ObjectName,            // const char* (ideally ustring)
+    AttributeName,         // const char* (ideally ustring)
+};
+
+// TODO: replace with std::variant
+class AttributeSpecArg {
+    union {
+        AttributeSpecBuiltinArg m_builtin;
+        int32_t m_int32;
+        float m_float;
+        uint64_t m_uint64;
+        void* m_ptr;
+    };
+    uint32_t m_type;
+
+public:
+    AttributeSpecArg() : m_type(0) {}
+    AttributeSpecArg(AttributeSpecBuiltinArg arg) : m_builtin(arg), m_type(1) {}
+    AttributeSpecArg(int32_t arg) : m_int32(arg), m_type(2) {}
+    AttributeSpecArg(float arg) : m_float(arg), m_type(3) {}
+    AttributeSpecArg(uint64_t arg) : m_uint64(arg), m_type(4) {}
+    AttributeSpecArg(void* arg) : m_ptr(arg), m_type(5) {}
+
+    template<typename T> bool is_holding() const
+    {
+        if (std::is_same<T, AttributeSpecBuiltinArg>::value)
+            return m_type == 1;
+        if (std::is_same<T, int32_t>::value)
+            return m_type == 2;
+        if (std::is_same<T, float>::value)
+            return m_type == 3;
+        if (std::is_same<T, uint64_t>::value)
+            return m_type == 4;
+        if (std::is_same<T, void*>::value)
+            return m_type == 5;
+        return false;
+    }
+
+    template<typename T> T get() const
+    {
+        assert(false);
+        return T();
+    }
+};
+
+template<>
+inline AttributeSpecBuiltinArg
+AttributeSpecArg::get() const
+{
+    assert(is_holding<AttributeSpecBuiltinArg>());
+    return m_builtin;
+}
+
+template<>
+inline int32_t
+AttributeSpecArg::get() const
+{
+    assert(is_holding<int32_t>());
+    return m_int32;
+}
+
+template<>
+inline float
+AttributeSpecArg::get() const
+{
+    assert(is_holding<float>());
+    return m_float;
+}
+
+template<>
+inline uint64_t
+AttributeSpecArg::get() const
+{
+    assert(is_holding<uint64_t>());
```

Comment:
Change asserts to OSL_DASSERT() or OSL_ASSERT depending on if you want them for debug only or not.